### PR TITLE
Add remote store protobuf/gRPC definitions

### PR DESCRIPTION
**Pull request by zehiko**
### What

In this PR we aim to define initial version of the V0 storage node APIs in the form of grpc protobuf definition. We also want to:
* settle on code generation
* type conversions (between rerun internal types and grpc types

(currently) Example usage:

```rust

use re_storage_types::v0::{
    storage_node_client::StorageNodeClient, RegisterRecordingsRequest
};

...

  let mut client = StorageNodeClient::connect("http://127.0.0.1:51234")
        .await
        .unwrap();

    let resp = client
        .register_recordings(RegisterRecordingsRequest {
            description: "test".to_string(),
            obj_storage:  ObjectStore {
              url: "dna.rd",
              bucket_name: "my_s3_bucket",
            },
            typ: re_storage_types::v0::RecordingType::Rrd.into(),
        })
        .await
        .unwrap()
        .into_inner();
```

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7634?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7634?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!
* [x] If have noted any breaking changes to the log API in `CHANGELOG.md` and the migration guide

- [PR Build Summary](https://build.rerun.io/pr/7634)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

File: crates/build/re_remote_store_types_builder/src/bin/build_re_remote_store_types.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    let workspace_dir = Utf8Path::new(env!("CARGO_MANIFEST_DIR"))
+        .parent()
+        .and_then(|p| p.parent())
+        .and_then(|p| p.parent())
+        .unwrap();
+
+    assert!(
+        workspace_dir.join("CODE_OF_CONDUCT.md").exists(),
+        "failed to find workspace root"
+    );
```
Comment (teh-cmc) [inline crates/build/re_remote_store_types_builder/src/bin/build_re_remote_store_types.rs:26]: Don't leave future reader extremely confused :no_mouth: 

```suggestion
    // Check for something that only exists in root:
    assert!(
        workspace_dir.join("CODE_OF_CONDUCT.md").exists(),
        "failed to find workspace root"
    );

File: crates/build/re_remote_store_types_builder/src/bin/build_re_remote_store_types.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    let definitions_dir_path = workspace_dir.join(PROTOBUF_DEFINITIONS_DIR_PATH);
+    let rust_generated_output_dir_path = workspace_dir.join(RUST_V0_OUTPUT_DIR_PATH);
+
+    re_log::info!("Running codegen for storage node types");
```
Comment (teh-cmc) [inline crates/build/re_remote_store_types_builder/src/bin/build_re_remote_store_types.rs:31]:
```suggestion
    re_log::info!(
        definitions=?definitions_dir_path,
        output=?rust_generated_output_dir_path,
        "Running codegen for storage node types",
    );

File: crates/store/re_remote_store_types/proto/rerun/v0/common.proto [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+message TimeColumnSelector {
+    Timeline timeline = 1;
+}
+
+message ComponentColumnSelector {
+    EntityPath entity_path = 1;
+    Component component = 2;
+    // TODO do we need join encoding?
```
Comment (teh-cmc) [inline crates/store/re_remote_store_types/proto/rerun/v0/common.proto:111]: Not for now at the very least -- we'll add it when/if we do.

File: crates/store/re_remote_store_types/proto/rerun/v0/common.proto [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    string id = 1;
+}
+
+// A recording can have multiple timelines, each is identified by a name, for example `log_tick`, `log_time`, etc.
+message Timeline {
+    string name = 1;
+}
+
+// A time range between start and end time points. Each 64 bit number can represent different time point data
+// depending on the timeline it is associated with.
```
Comment (teh-cmc) [inline crates/store/re_remote_store_types/proto/rerun/v0/common.proto:16]: This should specify that the range is inclusive etc.

But really, in general, everything in this file that matches an existing rust type should simply copy-paste the doc from the original rust types (they will likely grow into slightly different types over time and so will their docs and that's fine, but that's a good starting point).

File: crates/store/re_remote_store_types/proto/rerun/v0/remote_store.proto [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+package rerun.remote_store.v0;
+
+import "rerun/v0/common.proto";
+
+service StorageNode {
+    rpc ListRecordings(ListRecordingsRequest) returns (ListRecordingsResponse) {}
+    rpc Query(QueryRequest) returns (stream QueryResponse) {}
+    rpc GetRecordingMetadata(GetRecordingMetadataRequest) returns (GetRecordingMetadataResponse) {}
+    // TODO (zehiko) - should this be singular recording registration? Currently we can have 1 rrd => many recordings
```
Comment (teh-cmc) [inline crates/store/re_remote_store_types/proto/rerun/v0/remote_store.proto:11]: Here and everywhere, this should be `TODO(zehiko)`, not `TODO (zehiko)`.
That's important for all of our TODO-based automated tools to work properly.

The fact that this wasn't caught by the CI is a bug -- I'll make a separate PR for that.

File: crates/store/re_remote_store_types/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+//! necessary conversion code (in the form of `From` and `TryFrom` traits) in this crate.
+//!
+
+// Ignoring all warnings for the auto-generated code.
+#![allow(clippy::doc_markdown)]
+#![allow(clippy::derive_partial_eq_without_eq)]
+#![allow(clippy::enum_variant_names)]
+#![allow(clippy::unwrap_used)]
+#![allow(clippy::wildcard_imports)]
+#![allow(clippy::manual_is_variant_and)]
```
Comment (teh-cmc) [inline crates/store/re_remote_store_types/src/lib.rs:15]: You should be able to apply these only to the `mod _v0` below.

File: crates/store/re_remote_store_types/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+    impl From<re_log_types::ResolvedTimeRange> for TimeRange {
+        fn from(time_range: re_log_types::ResolvedTimeRange) -> Self {
+            Self {
+                start: time_range.min().as_i64(),
+                end: time_range.max().as_i64(),
+            }
+        }
+    }
+
+    impl TryFrom<Query> for re_dataframe::external::re_chunk_store::QueryExpression {
```
Comment (teh-cmc) [inline crates/store/re_remote_store_types/src/lib.rs:42]: Here and everywhere you shouldn't need `re_dataframe::external::re_chunk_store::QueryExpression`, simply `re_dataframe::QueryExpression`.

If that doesn't work then we have a re-export problem, and it's important to be fixed as these are supposed to be sane-to-use public APIs.

File: crates/store/re_remote_store_types/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+                .name;
+
+            #[allow(clippy::match_same_arms)]
+            let timeline = match timeline_name.as_str() {
+                "log_time" => Self::new_temporal(timeline_name),
+                "log_tick" => Self::new_sequence(timeline_name),
+                "frame" => Self::new_sequence(timeline_name),
+                "frame_nr" => Self::new_sequence(timeline_name),
+                _ => Self::new_temporal(timeline_name),
+            };
```
Comment (teh-cmc) [inline crates/store/re_remote_store_types/src/lib.rs:133]: Oof, that's a huge issue.

This needs a `TODO(cmc): QueryExpression::filtered_index gotta be a selector`.

File: crates/store/re_remote_store_types/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+                    entity_path: re_log_types::EntityPath::from("/somepath/c"),
+                    component: re_dataframe::external::re_chunk::ComponentName::new("component"),
+                    join_encoding: re_dataframe::external::re_chunk_store::JoinEncoding::default(),
+                }
+                .into(),
+            ]),
+        };
+
+        let query_expression: re_dataframe::external::re_chunk_store::QueryExpression =
+            query.try_into().unwrap();
```
Comment (teh-cmc) [inline crates/store/re_remote_store_types/src/lib.rs:335]: A full roundtrip would be nice:
```
let query_grpc_before = /* ... */;
let query_native = query_grpc_before.try_into();
let query_grpc_after =  query_native.into();
assert_eq!(query_grpc_before, query_grpc_after);
```
at which point you can probably remove the handwritten expectations

**Review by teh-cmc — CHANGES_REQUESTED** [review]
Nice! A few last things but we're very close.

File: crates/build/README.md
[the remark this answers is not in the document]
Comment (zehiko) [reply crates/build/README.md:1]: errr...I have no good explanation for this. But will certainly revert to old file. good catch!

File: crates/store/re_remote_store_types/proto/rerun/v0/common.proto (hunk 3 above)
Comment (zehiko) [reply crates/store/re_remote_store_types/proto/rerun/v0/common.proto:111]: thanks, removing the TODO

File: crates/store/re_remote_store_types/proto/rerun/v0/remote_store.proto (hunk 5 above)
Comment (zehiko) [reply crates/store/re_remote_store_types/proto/rerun/v0/remote_store.proto:11]: got it, tnx, fixed.

File: crates/store/re_remote_store_types/src/lib.rs (hunk 6 above)
Comment (zehiko) [reply crates/store/re_remote_store_types/src/lib.rs:15]: makes sense, updated

File: crates/store/re_remote_store_types/src/lib.rs (hunk 9 above)
Comment (zehiko) [reply crates/store/re_remote_store_types/src/lib.rs]: that makes sense, will add the other way conversion and update the unit test

File: crates/store/re_remote_store_types/src/lib.rs (hunk 7 above)
Comment (zehiko) [reply crates/store/re_remote_store_types/src/lib.rs:42]: missing exports addressed via https://github.com/rerun-io/rerun/pull/7709

File: crates/store/re_remote_store_types/proto/rerun/v0/common.proto (hunk 4 above)
Comment (zehiko) [reply crates/store/re_remote_store_types/proto/rerun/v0/common.proto:16]: I somehow missed this comment. Thanks, I've applied 95% of original documentation i.e. whenever it was actually there and made sense to be applied exactly to protobuf docs.

**Review by teh-cmc — APPROVED** [review]

**Comment by teh-cmc** [comment]
TODO:
* https://github.com/rerun-io/rerun/issues/7838

**Pull request merged**